### PR TITLE
Problem rework and subsampling

### PR DESCRIPTION
**Pull request by AlCap23**
Adds some more insights into the problem via `show` and `print`. Allows for naming a problem.
Assert if a problem is applicable to a basis. 

TODO: Subsampling of the problem

### CODE REVIEW

**Comment by AlCap23** [comment]
Current example 

```Julia
using DataDrivenDiffEq
using LinearAlgebra
using ModelingToolkit
using OrdinaryDiffEq
using Plots 
using Statistics
using StatsBase
using Random

@variables u[1:2]
basis = Basis([polynomial_basis(u, 5); sin.(u); cos.(u)], u)

function pendulum(u, p, t)
    x = u[2]
    y = -9.81sin(u[1]) - 0.1u[2]^3 - 0.2 * cos(u[1])
    return [x;y]
end

u0 = [0.99π; -1.0]
tspan = (0.0, 20.0)
dt = 0.1
prob = ODEProblem(pendulum, u0, tspan)
sol = solve(prob, Tsit5(), saveat=dt)

X = Array(sol)
t = sol.t
Random.seed!(1234)
X = X .+ 1e-1*randn(size(X))

dd_prob_noisy = ContinuousDataDrivenProblem(
    X, t, GaussianKernel()
    )


sampler = DataSampler(Split(ratio = 0.8), Batcher(n = 4, shuffle = true, repeated = true, batchsize_min = 40))
opt = ADMM(1e-3:1e-3:5e-1)
sol = solve(dd_prob_noisy, basis, opt, maxiter = 1000, by = :stat, progress = true, denoise = true, normalize = true, sampler = sampler)
sol
```

Output of the result and parameters

```julia
julia> println(result(sol))
Model ##Basis#413 with 2 equations
States : u[1] u[2]
Parameters : 7
Independent variable: t
Equations
Differential(t)(u[1]) = p₁*u[2]
Differential(t)(u[2]) = p₂*u[2] + p₃*(u[2]^3) + p₄*(u[2]^5) + p₅*sin(u[1]) + p₆*cos(u[1]) + p₇*cos(u[2])

julia> parameters(sol)
7-element Vector{Measurements.Measurement{Float64}}:
   0.958 ± 0.051
  -0.075 ± 0.096
  -0.072 ± 0.049
 -0.0025 ± 0.005
   -8.76 ± 0.31
    -0.1 ± 0.2
    -0.2 ± 0.14
```

The `DataSampler` consists of two major components which can be used individually. `Split` defines a train / test split.
`Batcher` creates data batches of the training data. Both take in random seeds ( need to expose this in the API ) to ensure reproducibility.

The keyword `by` can be set to use either 
 + :min which returns the best test error
 + :kfold which returns the best average error over all training sets
 + :stat which returns the mean +- std of all 

or anything which defaults to :min.

Additionally, the original output of SINDy now contains all the best sparsity knobs, which can be retrieved via `output(sol).\lambda`

File: src/problem/type.jl [hunk 1]
```diff
@@ -100,34 +100,40 @@ struct DataDrivenProblem{dType, cType, probType} <: AbstractDataDrivenProblem{dT
     """Input measurements (optional); Used for non-autonoumous problems"""
     U::AbstractMatrix{dType}
 
-
     """Parameters associated with the problem (optional)"""
     p::AbstractVector{dType}
+
+    """Name of the problem"""
+    name::String
```
Comment (ChrisRackauckas) [inline src/problem/type.jl:107]:
```suggestion
    name::Symbol
```

File: src/problem/type.jl [hunk 2]
```diff
@@ -220,22 +241,111 @@ A direct `DataDrivenProblem` useable for problems of the form `f(x,p,t,u) ↦ y`
...
+function ModelingToolkit.get_dvs(p::AbstractDiscreteProb, i = :, j = :)
+    ModelingToolkit.states(p, i, j)
+end
+
+function ModelingToolkit.observed(p::AbstractDataDrivenProblem, i = :, j = :)
+    x = getfield(p, :Y)
+    isempty(x) ? x : getindex(x, i, j)
+end
+
+function ModelingToolkit.controls(p::AbstractDataDrivenProblem, i = :, j = :)
```
Comment (ChrisRackauckas) [inline src/problem/type.jl:301]: this should be the inputs?

File: test/quicktest.jl [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+
+
+X = Array(sol)
+t = sol.t
+Random.seed!(1234)
+X = X .+ 1e-1*randn(size(X))
+
+dd_prob_noisy = ContinuousDataDrivenProblem(
+    X, t, GaussianKernel()
```
Comment (ChrisRackauckas) [inline test/quicktest.jl:35]: where is the kernel used?

File: src/problem/type.jl (hunk 2 above)
Comment (AlCap23) [reply src/problem/type.jl:301]: Yes, the external controls.

File: test/quicktest.jl (hunk 3 above)
Comment (AlCap23) [reply test/quicktest.jl:35]: In `src/problem/types.jl` L 206 / 227 . 

```Julia
function ContinuousDataDrivenProblem(X::AbstractMatrix, t::AbstractVector, collocation = InterpolationMethod(); kwargs...)
```

Just a copy paste from the DiffEqFlux collocation stuff + DataInterpolations

Comment (AlCap23) [reply test/quicktest.jl:35]: This is my local testfile btw, so feel free to ignore.

Comment (ChrisRackauckas) [reply test/quicktest.jl:35]: > Just a copy paste from the DiffEqFlux collocation stuff + DataInterpolations

Ahh I see, makes sense.

**Comment by AlCap23** [comment]
I still need to adapt the docs, but tests are green locally.

**Comment by AlCap23** [comment]
This is the same script. 

Screenshot 2022-02-04 at 17 34 31

**Comment by AlCap23** [comment]
I've flagged the test broken for now. There is no eminent reason for me why the result is good on `v.1.7` and off on `v.1.6`. 
But I'll have a look into this when this is partially done. But maybe we should not release for now ❓

**Pull request merged**